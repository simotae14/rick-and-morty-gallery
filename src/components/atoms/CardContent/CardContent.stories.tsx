import type { Meta, StoryObj } from '@storybook/react';

import CardContent from './index';

const meta = {
  title: 'Atoms/CardContent',
  component: CardContent,
  tags: ['autodocs'],
  argTypes: {
    name: {
      control: {
        type: 'text',
      },
    },
    image: {
      control: {
        type: 'text',
      },
    },
    status: {
      control: {
        type: 'text',
      },
    },
    species: {
      control: {
        type: 'text',
      },
    },
  },
  // The default value of the name arg to all stories
  args: { name: 'No Character' },
} satisfies Meta<typeof CardContent>;

export default meta;

type Story = StoryObj<typeof meta>;

export const EmptyProps: Story = {
  args: {
    name: undefined,
  },
};

export const WithProps: Story = {
  args: {
    name: 'Jerry Smith',
    image: 'https://rickandmortyapi.com/api/character/avatar/5.jpeg',
    status: 'Alive',
    species: 'Human',
  },
};
