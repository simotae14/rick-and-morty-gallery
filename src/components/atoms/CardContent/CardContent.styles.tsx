/* eslint-disable no-nested-ternary */
import styled from 'styled-components';

export const CardContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: inherit;
  padding: 0 16px;
  &:hover {
    cursor: pointer;
  }
`;

export const Image = styled.img`
  width: 140px;
  height: 140px;
  border-radius: 50%;
`;

export const Title = styled.h3`
  font-weight: 600;
  font-size: 1rem;
  margin: 0;
  width: 100%;
  text-align: center;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const SubTitle = styled.p`
  font-weight: 400;
  font-size: 0.75rem;
  color: #121c33;
  opacity: 0.6;
  margin: 0;
  width: 100%;
  text-align: center;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const StatusIcon = styled.span<{ status?: string }>`
  height: 0.5rem;
  width: 0.5rem;
  background-color: ${(props) =>
    props?.status === 'Alive'
      ? 'rgb(85, 204, 68)'
      : props?.status === 'Dead'
      ? 'rgb(214, 61, 46)'
      : 'rgb(158, 158, 158)'};
  border-radius: 50%;
  display: inline-block;
  margin-right: 0.375rem;
`;
