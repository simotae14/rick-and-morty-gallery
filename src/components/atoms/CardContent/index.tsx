import {
  CardContainer,
  Image,
  Title,
  SubTitle,
  StatusIcon,
} from './CardContent.styles.jsx';

import { CharacterSmall } from '../../../types/Character.js';

/**
 * Component that contains the information of a Character
 * inside the Gallery of characters
 */
function CardContent({
  name = 'No Character',
  image,
  status,
  species,
}: Partial<Omit<CharacterSmall, 'id'>>) {
  return (
    <CardContainer>
      <Image
        width="140"
        alt={name}
        src={image || '../../../assets/rick-morty.svg'}
      />
      <Title>{name}</Title>
      <SubTitle>
        <StatusIcon status={status} />
        {status} - {species}
      </SubTitle>
    </CardContainer>
  );
}

export default CardContent;
