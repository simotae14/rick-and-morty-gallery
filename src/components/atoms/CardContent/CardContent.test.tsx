/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable import/no-extraneous-dependencies */
import { describe, it } from 'vitest';
// eslint-disable-next-line import/no-extraneous-dependencies
import { screen, render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';

import CardContent from './index';

expect.extend(toHaveNoViolations);

const mockProps = {
  name: 'I am a title',
  image: 'http://image',
  status: 'Alive',
  species: 'Human',
};

const mockDeadCharacterProps = {
  name: 'I am a title',
  image: 'http://image',
  status: 'Dead',
  species: 'Human',
};

describe('Components => atoms => CardContent', () => {
  it('should render the CardContent with the default when no props are passed', () => {
    const { asFragment } = render(<CardContent />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not show the subtitle in CardContent when no props are passed', () => {
    render(<CardContent />);
    const subtitle = screen.queryByText('Alive - Human');
    expect(subtitle?.textContent).not.toBe(/Alive - Human/i);
  });

  it('should render the CardContent with some props passed', () => {
    const { asFragment } = render(<CardContent {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the CardContent with some props passed as dead character', () => {
    const { asFragment } = render(<CardContent {...mockDeadCharacterProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the CardContent title when is passed as prop', () => {
    render(<CardContent {...mockProps} />);
    const h3 = screen.getByRole('heading', { level: 3 });
    expect(h3.textContent).not.toBe(/i am a title/i);
  });

  it('should show the subtitle in CardContent when it is passed as prop', () => {
    render(<CardContent {...mockProps} />);
    const subtitle = screen.queryByText('Alive - Human');
    expect(subtitle?.textContent).toBe('Alive - Human');
  });

  it('should show the image in CardContent when it is passed as prop', () => {
    render(<CardContent {...mockProps} />);
    const image = screen.getByAltText('I am a title') as HTMLImageElement;
    expect(image.alt).toContain('I am a title');
  });

  it('should not fail any accessibility tests the CardContent without props', async () => {
    const { container } = render(<CardContent />);
    expect(await axe(container)).toHaveNoViolations();
  });

  it('should not fail any accessibility tests the CardContent with props', async () => {
    const { container } = render(<CardContent {...mockProps} />);
    expect(await axe(container)).toHaveNoViolations();
  });
});
