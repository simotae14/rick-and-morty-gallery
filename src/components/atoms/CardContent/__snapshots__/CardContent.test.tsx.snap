// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components => atoms => CardContent > should render the CardContent with some props passed 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir ca-DzqN"
  >
    <img
      alt="I am a title"
      class="sc-HzRMQ bhPYqq"
      src="http://image"
      width="140"
    />
    <h3
      class="sc-hIFLYB FsSHD"
    >
      I am a title
    </h3>
    <p
      class="sc-buuUZy dqwIcC"
    >
      <span
        class="sc-ivAPlj krxMbp"
        status="Alive"
      />
      Alive - Human
    </p>
  </div>
</DocumentFragment>
`;

exports[`Components => atoms => CardContent > should render the CardContent with some props passed as dead character 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir ca-DzqN"
  >
    <img
      alt="I am a title"
      class="sc-HzRMQ bhPYqq"
      src="http://image"
      width="140"
    />
    <h3
      class="sc-hIFLYB FsSHD"
    >
      I am a title
    </h3>
    <p
      class="sc-buuUZy dqwIcC"
    >
      <span
        class="sc-ivAPlj iNVWyK"
        status="Dead"
      />
      Dead - Human
    </p>
  </div>
</DocumentFragment>
`;

exports[`Components => atoms => CardContent > should render the CardContent with the default when no props are passed 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir ca-DzqN"
  >
    <img
      alt="No Character"
      class="sc-HzRMQ bhPYqq"
      src="../../../assets/rick-morty.svg"
      width="140"
    />
    <h3
      class="sc-hIFLYB FsSHD"
    >
      No Character
    </h3>
    <p
      class="sc-buuUZy dqwIcC"
    >
      <span
        class="sc-ivAPlj fpZVo"
      />
       - 
    </p>
  </div>
</DocumentFragment>
`;
